Add max retries per call, defaulting to 3

mjesun opened the pull request:
One of the requirements of the API was that we needed a way of ending childs that unexpectedly leave. We introduce the `maxRetries` variable; which measures how many times a child can be restarted (per call), before emitting an `error` event.

Tests were added so that a 100% coverage is still ensured on the module, as it was before. Also, `WorkerOptions` type was made stricter by making required the `forkOptions` property (which was anyway always passed).

SimenB commented on packages/jest-worker/src/__tests__/worker.test.js:
```diff
@@ -72,8 +76,35 @@ it('initializes the child process with the given workerPath', () => {
+  forkInterface.emit('exit');
+
+  // 3 retries + the initial load = 4.
+  expect(childProcess.fork.mock.calls.length).toBe(4);
```
` expect(childProcess.fork).toHaveBeenCalledTimes(4);`?

cpojer commented on packages/jest-worker/src/__tests__/worker.test.js:
```diff
@@ -23,16 +23,17 @@ let forkInterface;
-  childProcess.fork.mockImplementation(() => forkInterface);
+  childProcess.fork.mockImplementation(
+    () =>
+      (forkInterface = Object.assign(new EventEmitter(), {
```
can I buy curly braces and a few lines of code to make this more readable, please? :)

mjesun replied:
Give thanks to prettier which made it unreadable :)

cpojer commented on packages/jest-worker/src/index.js:
```diff
@@ -68,6 +68,7 @@ export default class {
     // Build the options once for all workers to avoid allocating extra objects.
     const workerOptions = {
       forkOptions: options.forkOptions || {},
+      maxRetries: options.maxRetries || Infinity,
```
I'm not sure Infinity is a sensible default. Why would you ever want to retry more than say 3 or 5 times? Can we set it to 3 or 5? People can still pass "Infinity" if they want, but the default should be to crash after a few retries.

cpojer commented on packages/jest-worker/src/worker.js:
```diff
@@ -45,15 +46,19 @@ import type {
+
     this._options = options;
     this._queue = [];
+    this._startedTimes = 0;
```
_retries?

mjesun commented:
>  With this change, you explicitly have to listen to error events.

I'm against crashing on the `Promise` level, because it would make confusing to know whether the error came from the called function, or the farm (i.e. user VS infrastructure level).

Triggering an `error` event is the standard and accepted way of dealing with this kind of stuff in Node, since an unlistened `error` event makes the process crash, like a throw; and it will not hide errors within call returns.

> Also, I don't quite understand the code in this change. What is responsible for the restarts of a specific worker?

The restart of a specific worker is handled by the `_exit` method; that checks the exit code, and if it was not zero, it calls `_initialize` back. The `_exit` method is the callback for the `exit` event, triggered by the `ChildProcess` object. That logic existed earlier.

cpojer commented:
Got it, thanks for the explanation. I guess before and after this PR, with default settings, doesn't change any behavior, which is why I missed this.

I'm not sure I agree about the error handling. When a process crashes, what happens when the retries for that same call fail three times (if there are three retries)? Doesn't that promise reject with the process terminated error? If yes, that should be enough to handle the failure case.

Also, I think I should have been more clear about what I expect `jest-worker` to do when things go south. When a worker crashes, I would like us to spawn another worker and apply the retries to that same call N times (as specified in the options) but, assuming that one completes within the specified amount of retries, future calls use their own retry counter. In short, the `maxRetries` option should be applied to individual calls to jest-worker, and not count towards all calls over the lifetime of one of the workers. This allows jest-worker to tape over a bunch of intermittent failures throughout its lifetime rather than crashing when hitting a limit, which will undoubtedly happen at some point during the lifetime of the app. The goal here is to sensible retry and see if we can keep running without crashing the app. Does that make sense?

mjesun commented:
@cpojer Gotcha, now I understood :) I modified the diff so that it matches your point.

cpojer commented on packages/jest-worker/README.md:
```diff
@@ -59,6 +59,10 @@ List of method names that can be called on the child processes from the parent p
 
+#### `maxRetries: number` (optional)
+
+Maximum amount of times that a dead child can be re-spawned, per call. Defaults to `5`, pass `Infinity` to allow endless retries.
```
Let's do 3 as a default. If something crashes the process three times in a row, it's probably broken.

cpojer commented on packages/jest-worker/src/worker.js:
```diff
@@ -91,8 +92,24 @@ export default class {
+    // coming from the child. This avoids code duplication related with cleaning
+    // the queue, and scheduling the next call.
+    if (this._retries > this._options.maxRetries) {
+      const error = new Error('Call retries were exceeded');
```
Could this be a subclass of error with a `.type` field or something so that people can distinguish it from user errors?

cpojer commented:
Nice! Thanks for updating it, this makes more sense to me. Just to confirm again, if a call crashes the process, we will restart that process and issue the same call again using the same worker class instance that also keeps track of the number of retries?

mjesun commented:
Addressed all the feedback; decreased retries to `3` and added the extra `type` error.

> [...] if a call crashes the process, we will restart that process and issue the same call again using the same worker class instance that also keeps track of the number of retries?

Yes 🙂 The code loops between `initialize` and `exit` up until the counter goes above the limit; then it emits a fake reply through `_receive` with the error. This, in turn, causes the held promise to be rejected, and makes `_process` pick the next item in the queue.
